src/business_logic/transaction/internal_objects.rs

```diff
@@ -166,3 +179,370 @@ impl InternalDeploy {
+        let fee_transfer_info =
+            execute_fee_transfer(state, general_config, tx_context, actual_fee)?;
+
+        Ok((Some(fee_transfer_info), actual_fee))
+    }
+
+    fn handle_nonce(
+        &self,
+        state: &mut CachedState<InMemoryStateReader>,
+    ) -> Result<(), TransactionError> {
+        if self.version > 0x8000_0000_0000_0000 {
+            return Err(TransactionError::StarknetError(
+                "Don't handle nonce for version 0".to_string(),
+            ));
+        }
+
+        let contract_address = self.account_contract_address();
+        let current_nonce = state.get_nonce_at(&contract_address)?.to_owned();
+        if current_nonce != self.nonce {
+            return Err(TransactionError::InvalidTransactionNonce(
+                current_nonce.to_string(),
+                self.nonce.clone().to_string(),
+            ));
+        }
+
+        state.increment_nonce(&contract_address)?;
+
+        Ok(())
+    }
+
+    pub fn apply_specific_sequential_changes(
+        &self,
+        state: &mut CachedState<InMemoryStateReader>,
+        general_config: StarknetGeneralConfig,
+        actual_resources: HashMap<String, Felt>,
+    ) -> Result<FeeInfo, TransactionError> {
+        self.handle_nonce(state)?;
+
+        self.charge_fee(state, actual_resources, general_config)
+    }
+}
+
+// ---------------
+//     Tests
+// ---------------
+
+#[cfg(test)]
+mod tests {
+    use felt::Felt;
+    use num_traits::Num;
+    use std::{collections::HashMap, path::PathBuf};
+
+    use crate::{
+        business_logic::{
+            execution::objects::{CallInfo, CallType, TransactionExecutionInfo},
+            fact_state::{
+                contract_state::ContractState, in_memory_state_reader::InMemoryStateReader,
+            },
+            state::cached_state::CachedState,
+        },
+        definitions::{
+            general_config::{StarknetChainId, StarknetGeneralConfig},
+            transaction_type::TransactionType,
+        },
+        services::api::contract_class::{ContractClass, EntryPointType},
+        starknet_storage::dict_storage::DictStorage,
+        utils::Address,
+    };
+
+    use super::InternalDeclare;
+
+    #[test]
+    fn test_internal_declare() {
+        let path = PathBuf::from("starknet_programs/account_without_validation.json");
+        let contract_class = ContractClass::try_from(path).unwrap();
+        let chain_id = StarknetChainId::TestNet.to_felt();
+
+        let internal_declare = InternalDeclare::new(
+            contract_class.clone(),
+            chain_id,
+            Address(1.into()),
+            0,
+            0,
+            Vec::new(),
+            0.into(),
+        )
+        .unwrap();
+
+        // Instantiate CachedState
+        let mut contract_class_cache = HashMap::new();
+
+        //  ------------ contract data --------------------
+
+        let address = Address(1.into());
+        let class_hash = [1; 32];
```

Comment:
The class hash of this contract class is given by compute_class_hash. Specifically the field class hash of InternalDeclare